docs: template registration UTXO (#2)
tari-project/rfcs: src/RFC-0306_DANTemplateRegistration.md

assigned a contract via the [contract definition], the [validator node] performs the following actions to obtain a template:

1. It scans the blockchain for a template matching the `binary_checksum`;
2. it downloads the WASM binary and verifies the checksum and stores the binary and associated registration [UTXO] metadata;
3. a validator node operator may choose to build the binary from source as per `BuildInfo` and use the resulting binary;

In the event that a new validator is added to [contract constitution], but the original template registration is unavailable,
the [validator node] SHOULD make the template available to new committee members. New committee members SHOULD confirm the checksum
with a 2/3 majority of the validator committee to ensure that the correct copy is received.

## Spending the Template Registration UTXO

These cases apply to spending the template registration [UTXO]:
1. The author MAY spend into another template registration [UTXO]
* This effectively withdraws (yanks) the previous version of the template.
* The template name and author SHOULD be identical. This MAY be enforced by covenant.

> * The template version SHOULD be incremented.

Reviewer: ```suggestion
* The template version MUST be incremented.
```

> * Any live contracts SHOULD upgrade their definitions to run the new template.

Reviewer: ```suggestion
* It is RECOMMENDED that live contracts upgrade their definitions to run the new template.
```

| binary_checksum   | SHA2 checksum      | A SHA2 checksum of the WASM binary.                                                                                                   |
| binary_address    | [Multiaddr]        | A [multiaddr] pointing to the WASM module binary download. This may be an HTTP, ONION, p2p, or IPFS address. Maximum byte length: 255 |

The `author_signature` is a Schnorr proof that commits to the template fields contained in the [UTXO], namely
`template_name`, `template_version`, `build_info`, `binary_checksum`, and `binary_address`.

The base node acts as a notary for this data, it is not responsible for the validity of the template fields. However, it must
not allow malformed/invalid data to be committed to the blockchain.

Therefore, some additional base-layer consensus rules are required for a `TEMPLATE_REGISTRATION` [UTXO]:
* A base node MUST validate the `author_signature`; and

> * The [UTXO] MUST have a relative time-lock of 100 blocks.

Reviewer 2: What's the reason for the lock height?
I can think of reasons why you might want to push an updated version shortly. For example, you realized there is a bug in the just pushed version, now you have to wait for 200 minutes before you can push your fix?

Author: Yeah this was chosen pretty arbitrarily. I think it's more around having some assurance for this UTXO existing on the blockchain when referencing this template in a contract - which this doesn't really answer so I'll remove this.

Will have to put more thought into how we reference template UTXOs or if we copy the template data into the contract definition/constitution which is kind of bloaty.

the [validator node] SHOULD make the template available to new committee members. New committee members SHOULD confirm the checksum
with a 2/3 majority of the validator committee to ensure that the correct copy is received.

## Spending the Template Registration UTXO

These cases apply to spending the template registration [UTXO]:
1. The author MAY spend into another template registration [UTXO]
* This effectively withdraws (yanks) the previous version of the template.
* The template name and author SHOULD be identical. This MAY be enforced by covenant.
* The template version SHOULD be incremented.
* Any live contracts SHOULD upgrade their definitions to run the new template.
2. The author MAY spend to a "vanilla" [UTXO] to reclaim their deposit.
* This effectively withdraws (yanks) the template.
* The template MAY remain in use on existing contracts. In fact, anyone may now re-register the template.

Alternatives:
* Consensus rule that prevents spending of the template while used by other contracts
* The contract definition may have copy the template binary_checksum etc. Validator nodes may mirror the code for new VNs

> ## Upgrading a Template

Reviewer 2: We should spec this out more as this could be troublesome if they upgrade out of sync
